### PR TITLE
SONARJAVA-1996 Report flows when invocation leads to exception

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java [hunk 1]
```diff
@@ -138,6 +148,28 @@ public int hashCode() {
...
   }
 
+  public Set<List<JavaFileScannerContext.Location>> exceptionFlows() {
+    Set<List<JavaFileScannerContext.Location>> flows = FlowComputation.flow(node.parent(), svCausingException, domains(node.programState.getConstraints(svCausingException)));
+    Tree syntaxTree = node.programPoint.syntaxTree();
+    ImmutableSet.Builder<List<JavaFileScannerContext.Location>> flowBuilder = ImmutableSet.builder();
+
+    for (List<JavaFileScannerContext.Location> flow : flows) {
+      List<JavaFileScannerContext.Location> newFlow = ImmutableList.<JavaFileScannerContext.Location>builder()
+        .add(new JavaFileScannerContext.Location(exceptionType().name() + " is thrown.", syntaxTree))
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java]: MAJOR NullPointerException might be thrown as 'exceptionType' is nullable here [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java (hunk 1 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java]: Cannot happen, exception has to be provided to create a `ExceptionalCheckBasedYield`.

I don't know if I should override the method and annotate it with `@NonNull`, but I will then have an issue from [S2637](https://jira.sonarsource.com/browse/RSPEC-2637) saying that the method is nonnull and returning null if I call super.exceptionType() within its body.

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java]: Wouldn't the correct fix actually be able to get rid of this null by having unknownType as default value ?

Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java]: Not sure, the `null` exception currently relates to any runtime exception which can happen when invoking the method. Another approach could be to chose to throw a (typed) runtime exception, as we are now able to load the type.

Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/ExceptionalCheckBasedYield.java]: But definitely out of the scope of this ticket.

**Pull request merged**